text/0028-bee-storage/0022-bee-storage.md

```diff
@@ -0,0 +1,212 @@
++ Feature name: `bee-storage`
++ Start date: 2019-13-11
++ RFC PR: [iotaledger/bee-rfcs#28](https://github.com/iotaledger/bee-rfcs/pull/)
++ Bee issue:  [iotaledger/bee#65](https://github.com/iotaledger/bee/issues/65)
+
+# Summary
+
+An abstraction for the persistent module which will be exclusively used by the Tangle module
+for CRUD operations 
+
+# Motivation
+
+The Tangle is a DAG where each vertex's data is a transaction,
+this data should of course be persisted, it is better to have this logic
+and functionality grouped into one separate module that exposes a clear API
+rather than have it in several places, this will also allow for implementing different database backends
+that will implement the storage API
+
+Whenever a bee module requires updating a transaction, it will do so by interacting with the Tangle module
+which will internally use this storage module
+
+use cases:
+
+1.When a new transaction arrives to the node and is not in the database, we would want to insert it as a new record
+2.When a transaction is required we would want to retrieve it from the database
+3.When we want to update a transaction as being solid
+4.When we want to delete a transaction during the local snapshot process
+
+
+# Detailed design
+
+```rust
+
+use std::{collections::HashMap, collections::HashSet, rc::Rc};
+
+/// A transaction hash. To be replaced later with whatever implementation is required.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub struct TxHash(u64);
+impl TxHash {
+    pub fn is_genesis(&self) -> bool {
+        unimplemented!()
+    }
+}
+
+/// A transaction address. To be replaced later with whatever implementation is required.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub struct TxAddress(u64);
+
+/// A transaction. Cannot be mutated once created.
+pub struct Tx {
+    hash: TxHash,
+    trunk: TxHash,
+    branch: TxHash,
+    body: (),
+}
+
+/// A milestone hash. To be replaced later with whatever implementation is required.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub struct MilestoneHash(u64);
+impl MilestoneHash {
+    pub fn is_genesis(&self) -> bool {
```

Comment:
I would be helpful to also add doc comment on what this is used for.

Reply:
sure